docs/resources/samples/performance.md

```diff
@@ -0,0 +1,151 @@
+---
+title: 'Performance optimization'
+description: 'Learn how to optimize read and write performance in Office Scripts.'
+ms.date: 03/03/2021
+localization_priority: Normal
+---
+
+# Read and write performance optimization
+
+## Basic perf optimization
+
+See the [getting started performance section](getting-started.md#basic-performance-considerations) for the basics of Office Scripts.
+
+## Large data write scenario
+
+The `setValues()` Range API allows setting the values of a range. This API has data limitations depending on various factors such as data size, network settings, etc. In order to reliably update a large range of data, you'll need to think about doing data updates in smaller chunks. This script attempts to do this and writes rows of a range in chunks so that if a large range needs to be updated, it can be done in smaller parts. **Warning**: It has not been tested across various sizes so be aware of that if you want to use this in your script. As we have opportunity to test, we'll update with findings around how it performs for various data sizes.
+
+This script selects 1K cells per chunk but you can override to test out how it works for you. It updates 100k rows with 6 columns of data. Run this on a blank sheet to examine.
+
+```ts
+function main(workbook: ExcelScript.Workbook) {
+  const sheet = workbook.getActiveWorksheet();
+
+  let data: (string | number | boolean)[][] = [];
+  // Number of rows in the random data (x 6 columns).
+  const sampleRows = 100000;
+
+  console.log(`Generating data...`)
+  // Dynamically generate some random data for testing purpose. 
+  for (let i = 0; i < sampleRows; i++) {
+    data.push([i, ...[getRandomString(5), getRandomString(20), getRandomString(10), Math.random()], "Sample data"]);
+  }
+
+  console.log(`Calling update range function...`);
+  const updated = updateRangeInChunks(sheet.getRange("B2"), data);
+  if (!updated) {
+    console.log(`Update did not take place or complete. Check and run again.`)
+  }
+
+  return;
+}
+
+function updateRangeInChunks(
+  startCell: ExcelScript.Range,
+  values: (string | boolean | number)[][],
+  cellsInChunk: number = 10000
+): boolean {
+
+  const startTime = new Date().getTime();
+  console.log(`Cells per chunk setting: ${cellsInChunk}`);
+  if (!values) {
+    console.log(`Invalid input values to update.`);
+    return false;
+  }
+  if (values.length === 0 || values[0].length === 0) {
+    console.log(`Empty data -- nothing to update.`);
+    return true;
+  }
+  const totalCells = values.length * values[0].length;
+
+  console.log(`Total cells to update in the target range: ${totalCells}`);
+  if (totalCells <= cellsInChunk) {
+    console.log(`No need to chunk -- updating directly`);
+    updateTargetRange(startCell, values);
+    return true;
+  }
```

Comment:
Once again, there's a lot of preamble. Let's consider streamlining the experience instead of encouraging customers to change values.

Reply:
Tracking with [feature](https://office.visualstudio.com/OC/_workitems/edit/4924924)